src/kzg10/subproductdomain.rs

```diff
@@ -0,0 +1,424 @@
+use crate::Error;
+use ark_ec::PairingEngine;
+use ark_ff::{FftField, Field, Zero};
+use ark_poly::{Polynomial, UVPolynomial};
+
+use ark_poly::polynomial::univariate::DensePolynomial as Poly;
+
+/// Where indicated, algorithms are from Modern Computer Algebra, 3rd edition, by Gathen and Gerhard
+/// Abbreviated as GG
+/// Let M(n) denote the time to multiply.
+
+/// GG Algorithm 9.3
+/// Computes the inverse of f mod x^l
+/// Takes O(M(l)) field arithmetic operations
+pub fn inverse_mod_xl<F: FftField>(f: &Poly<F>, l: usize) -> Option<Poly<F>> {
+    let r = ark_std::log2(l); // Compute ceil(log_2(l))
+    if let Some(f_0_inv) = f.coeffs[0].inverse() {
+        // Invert f(0)^-1 if possible
+        let mut g = Poly::<F> {
+            coeffs: vec![f_0_inv], // Constant polynomial f(0)^-1
+        };
+
+        let mut i = 2usize;
+        // Use Newton iteration which converges to the inverse mod x^l
+        for _ in 0..r {
+            g = &(&g + &g) - &(f * &(&g * &g)); //TODO: is g*2 better than g+g?
+            g.coeffs
+                .resize(ark_std::cmp::min(g.coeffs.len(), i), F::zero()); // Take g remainder mod x^{2^i}
+            i *= 2;
+        }
+        Some(g)
+    } else {
+        // No inverse exists because f(0)^-1 does not exist
+        None
+    }
+}
+
+/// GG chapter 9.1
+/// Computes the rev_m(f) function in place
+/// rev_m(f) = x^m f(1/x)
+pub fn rev<F: FftField>(f: &mut Poly<F>, m: usize) {
+    assert!(f.coeffs.len() - 1 <= m);
+    for _ in 0..(m - (f.coeffs.len() - 1)) {
+        f.coeffs.push(F::zero());
+    }
+    f.reverse();
+}
+
+/// GG Algorithm 9.5
+/// Divide f by g in nearly linear time
+pub fn fast_divide_monic<F: FftField>(f: &Poly<F>, g: &Poly<F>) -> (Poly<F>, Poly<F>) {
+    //assert_eq!(g.coeffs.last(), F::one()); //TODO: check monic condition
```

Comment:
Did you intend to uncomment this?

Reply:
oh, I did intend to revisit this later, because it seems that this could be handled in a more general way (for example, a MonicPolynomial type, I'm not sure) which provides compile time safety. But of course that's another entire project by itself.

Definitely for right now the assert should be uncommented as a runtime check

Reply:
```suggestion
    assert_eq!(g.coeffs.last(), F::one()); // check that `g` is monic
```